nip47/controllers/create_connection_controller.go

```diff
@@ -0,0 +1,135 @@
+package controllers
+
+import (
+	"context"
+	"slices"
+	"time"
+
+	"github.com/getAlby/hub/constants"
+	"github.com/getAlby/hub/logger"
+	"github.com/getAlby/hub/nip47/models"
+	"github.com/getAlby/hub/nip47/permissions"
+	"github.com/nbd-wtf/go-nostr"
+	"github.com/sirupsen/logrus"
+)
+
+type createConnectionParams struct {
+	Pubkey            string                 `json:"pubkey"` // pubkey of the app connection
+	Name              string                 `json:"name"`
+	RequestMethods    []string               `json:"request_methods"`
+	NotificationTypes []string               `json:"notification_types"`
+	MaxAmount         uint64                 `json:"max_amount"`
+	BudgetRenewal     string                 `json:"budget_renewal"`
+	ExpiresAt         *uint64                `json:"expires_at"` // unix timestamp
+	Isolated          bool                   `json:"isolated"`
+	Metadata          map[string]interface{} `json:"metadata,omitempty"`
+}
+
+type createConnectionResponse struct {
+	// pubkey is given, user requesting already knows relay.
+	WalletPubkey string `json:"wallet_pubkey"`
+}
+
+func (controller *nip47Controller) HandleCreateConnectionEvent(ctx context.Context, nip47Request *models.Request, requestEventId uint, publishResponse publishFunc) {
+	params := &createConnectionParams{}
+	resp := decodeRequest(nip47Request, params)
+	if resp != nil {
+		publishResponse(resp, nostr.Tags{})
+		return
+	}
+
+	logger.Logger.WithFields(logrus.Fields{
+		"request_event_id": requestEventId,
+		"params":           params,
+	}).Info("creating app")
+
+	var expiresAt *time.Time
+	if params.ExpiresAt != nil {
+		expiresAtUnsigned := *params.ExpiresAt
+		expiresAtValue := time.Unix(int64(expiresAtUnsigned), 0)
+		expiresAt = &expiresAtValue
+	}
+
+	maxAmountSat := params.MaxAmount / 1000
+
+	// explicitly do not allow creating an app with create_connection permission
+	if slices.Contains(params.RequestMethods, models.CREATE_CONNECTION_METHOD) {
+		publishResponse(&models.Response{
+			ResultType: nip47Request.Method,
+			Error: &models.Error{
+				Code:    constants.ERROR_INTERNAL,
+				Message: "cannot create a new app that has create_connection permission via NWC",
+			},
+		}, nostr.Tags{})
+		return
+	}
+
+	// ensure there is at least one request method
+	if len(params.RequestMethods) == 0 {
+		publishResponse(&models.Response{
+			ResultType: nip47Request.Method,
+			Error: &models.Error{
+				Code:    constants.ERROR_INTERNAL,
+				Message: "No methods provided",
```

Comment:
I'm assuming this is because otherwise the user will see `scopes` instead of `methods`. So should this be `request methods`